---
layout: lesson
root: .
---

[Last lesson](https://cloud-span.github.io/03genomics/) you performed quality control on your reads. This lesson we will align those reads to a reference genome, and end by identifying and visualizing variations among these samples. 

As you progress through this lesson, keep in mind that, even if you aren't going to be doing this same workflow in your research, you will be learning some very important lessons about using command-line bioinformatic tools. 
What you learn here will enable you to use a variety of bioinformatic tools with confidence and greatly enhance your research efficiency and productivity.


![variant calling](img\variant_calling.png)

> ## Prerequisites
>
> This lesson assumes a working understanding of the bash shell. If you haven't already completed [Prenomics](https://cloud-span.github.io/prenomics00-intro/), and aren't familiar with the bash shell, please review those materials
> before starting this lesson.
>
> This lesson also assumes some familiarity with biological concepts, including the structure of DNA, nucleotide abbreviations, and the concept of genomic variation within a population. 
>
> This lesson is part of a course that uses data hosted on an Amazon Machine Instance (AMI). Course participants will be given information on how to log-in to the AMI during the course. Information on preparing for the course is provided on the [Cloud-SPAN Genomics setup page](https://cloud-span.github.io/01genomics/setup.html).
{: .prereq}
